providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+        if (blake2b(outbuf, outlen_curr, inbuf, BLAKE2B_OUTBYTES, NULL, 0) != 1)
+            goto fail;
+        memcpy(out, outbuf, outlen_curr);
+    }
+    ret = 1;
+
+fail:
+    EVP_MAC_free_ctx(ctx);
+    EVP_MAC_free(mac);
+    return ret;
+}
+
+static void kdf_argon2_init(KDF_ARGON2 *ctx, ARGON2_TYPE type)
+{
+    ARGON2_CTX *c;
+
+    c = &ctx->ctx;
+    memset(c, 0, sizeof(*c));
+    c->outlen = 64;
+    c->t_cost = 3;
+    c->m_cost = ARGON2_MIN_MEMORY;
+    c->lanes = 1;
+    c->threads = 1;
+    c->flags = ARGON2_DEFAULT_FLAGS;
+    c->version = ARGON2_VERSION_NUMBER;
+    c->type = type;
+    c->openssl_ctx = PROV_LIBRARY_CONTEXT_OF(ctx->provctx);
+}
+
+static void *kdf_argon2d_new(void *provctx)
+{
+    KDF_ARGON2 *ctx;
+
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    if (ctx == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    ctx->provctx = provctx;
+
+    kdf_argon2_init(ctx, ARGON2_D);
+    return ctx;
+}
+
+static void *kdf_argon2i_new(void *provctx)
+{
+    KDF_ARGON2 *ctx;
+
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    if (ctx == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    ctx->provctx = provctx;
+
+    kdf_argon2_init(ctx, ARGON2_I);
+    return ctx;
+}
+
+static void *kdf_argon2id_new(void *provctx)
+{
+    KDF_ARGON2 *ctx;
+
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    if (ctx == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    ctx->provctx = provctx;
+
+    kdf_argon2_init(ctx, ARGON2_ID);
+    return ctx;
+}
+
+static void kdf_argon2_free(void *vctx)
+{
+    ARGON2_CTX *ctx;
+    KDF_ARGON2 *ctx_wrap;
+
+    ctx_wrap = (KDF_ARGON2 *) vctx;
+    ctx = &ctx_wrap->ctx;
+    if (ctx->out != NULL)
+        OPENSSL_clear_free(ctx->out, ctx->outlen);
+
+    if (ctx->pwd != NULL)
+        OPENSSL_clear_free(ctx->pwd, ctx->pwdlen);
+
+    if (ctx->salt != NULL)
+        OPENSSL_clear_free(ctx->salt, ctx->saltlen);
+
+    if (ctx->secret != NULL)
+        OPENSSL_clear_free(ctx->secret, ctx->secretlen);
+
+    if (ctx->ad != NULL)
+        OPENSSL_clear_free(ctx->ad, ctx->adlen);
+
+    memset(ctx, 0, sizeof(*ctx));
+    OPENSSL_free(ctx_wrap);
+}
+
+static int kdf_argon2_derive(void *vctx, unsigned char *out, size_t outlen)
+{
+    ARGON2_CTX *ctx;
+    KDF_ARGON2 *ctx_wrap;
+    uint32_t memory_blocks, segment_length;
+
+    ctx_wrap = (KDF_ARGON2 *)vctx;
+    ctx = &ctx_wrap->ctx;
+
+    if (ctx->pwd == NULL || ctx->pwdlen == 0) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_PASS);
+        return 0;
+    }
+
+    if (ctx->salt == NULL || ctx->saltlen == 0) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_SALT);
+        return 0;
+    }
+
+    if (outlen > UINT32_MAX)
+        return 0;
+
+    if (outlen == 0)
+        return 0;
+
+    if (outlen != ctx->outlen)
+        kdf_argon2_ctx_set_digest_length(ctx, (uint32_t) outlen);
+
```

Comment:
maybe use
```
sz = ctx->outlen + 1;
```
Then it can be used in the cleanup also.

Reply:
@ckalina @slontis is this comment still relevant?

Reply:
`ctx->outlen` is set to a default value of 64 during initialization and left unchanged provided `OSSL_KDF_PARAM_SIZE` isn't set.

This value is not the minimal supported, e.g., `outlen = 32` passed to derive is still ok as far as Argon2 is concerned. In this case, if `sz` means `outlen`, we might write past `out` in some cases.

I changed it a bit here: https://github.com/openssl/openssl/blob/b94350f527e540e6262069c16f09b8db6f0c6db8/providers/implementations/kdfs/argon2.c#L1108

@slontis Do you think there's anything to be done about this hunk?

Reply:
Unless there are objections, marking as resolved. Please feel free to unresolve if there's anything to do about this hunk.